apps/src/applab/designMode.js

```diff
@@ -1043,6 +1043,13 @@ designMode.onInsertEvent = function(code) {
   Applab.scrollToEnd();
 };
 
+designMode.onInsertAICode = function(code) {
+  Applab.appendToEditor(code);
+  getStore().dispatch(actions.changeInterfaceMode(ApplabInterfaceMode.DESIGN));
```

Comment:
interesting... do you know why we have to switch to design mode for this to work?

Reply:
This is my understanding, but I'm still not 100% clear on all the intricacies of App Lab... 

Design elements are typically created:  1.) by the user in design mode or 2.) as starter elements from the levelbuilder via the `start_html` property 

In [applab.js,](https://github.com/code-dot-org/code-dot-org/blob/d8fc97b757583e073a72fb82a17072eb631da901/apps/src/applab/applab.js#L317) we only get the HTML if designModeViz element is visible. 
```
Applab.getHtml = function() {
  // This method is called on autosave. If we're about to autosave, let's update
  // levelHtml to include our current state.
  if ($('#designModeViz').is(':visible')) {
    designMode.serializeToLevelHtml();
  }
  return Applab.levelHtml;
};
```

When there is starter HTML from levelbuilder we set the HTML [when the App is initialized](https://github.com/code-dot-org/code-dot-org/blob/d8fc97b757583e073a72fb82a17072eb631da901/apps/src/applab/applab.js#L548). 

In this case, I'm adding design elements outside the context of design mode and outside the starterHtml code path, so the easiest way to get those elements to be rendered is to briefly switch into design mode, making the designModeViz element visible, thus popping into the conditional that will set the the level html and render the autogenerated elements.

Reply:
ah, that makes sense. could you add a comment above this line that explains why we have to switch to design mode here?

Reply:
yep - will do!